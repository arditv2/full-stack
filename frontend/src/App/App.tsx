import React, { FC } from 'react';
import { connect } from 'react-redux';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Alert from 'react-bootstrap/Alert';
import Spinner from 'react-bootstrap/Spinner';

import { getPlacesAction } from './store/places/actions';
import { IPlace, ISearchForm } from './store/places/types';
import { RootReducerType } from '../configureStore';

import SearchForm from '../Components/SearchForm';
import PlacesList from '../Components/PlacesList';

import styles from './App.module.scss';

interface StateProps {
	loading: boolean;
	error?: string;
	places?: IPlace[];
}

interface DispatchProps {
	getPlaces: (payload: ISearchForm) => void;
}

interface OwnProps {}

type Props = StateProps & DispatchProps & OwnProps;

export const App: FC<Props> = ({ loading, error, places, getPlaces }) => (
	<Container fluid className='mb-3'>
		<Row>
			<Col md={{ span: 8, offset: 2 }} lg={{ span: 6, offset: 3 }}>
				{loading && (
					<Container fluid className={styles.spinner_container}>
						<Spinner animation='border' variant='primary' />
					</Container>
				)}
				{!loading && <SearchForm handleSubmit={getPlaces} />}
				{!loading && !error && places && Boolean(places.length) && (
					<PlacesList places={places} />
				)}
				{!loading && !error && places && Boolean(!places.length) && (
					<Alert variant='info'>
						There are no places related to your search
					</Alert>
				)}
				{!loading && error && <Alert variant='danger'>{error}</Alert>}
			</Col>
		</Row>
	</Container>
);

const mapState = (state: RootReducerType) => ({
	loading: state.places.loading,
	error: state.places.error,
	places: state.places.places,
});

const mapDispatch = {
	getPlaces: (payload: ISearchForm) => getPlacesAction(payload),
};

export default connect(mapState, mapDispatch)(App);
