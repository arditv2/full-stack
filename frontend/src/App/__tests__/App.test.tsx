import React from 'react';
import { shallow } from 'enzyme';

import { App } from '../App';

describe('App component', () => {
	const initialProps = {
		loading: false,
		getPlaces: jest.fn(),
	};

	it('renders only Spinner when loading', () => {
		const props = { ...initialProps, loading: true };
		const wrapper = shallow(<App {...props} />);
		expect(wrapper.find('Spinner')).toHaveLength(1);
		expect(wrapper.find('SearchForm')).toHaveLength(0);
		expect(wrapper.find('PlacesList')).toHaveLength(0);
		expect(wrapper.find('Alert')).toHaveLength(0);
		expect(wrapper).toMatchSnapshot();
	});

	it('renders SearchForm and Alert if there is an error', () => {
		const mockErrorMessage = 'Something failed';
		const props = { ...initialProps, error: mockErrorMessage };
		const wrapper = shallow(<App {...props} />);
		expect(wrapper.find('Spinner')).toHaveLength(0);
		expect(wrapper.find('SearchForm')).toHaveLength(1);
		expect(wrapper.find('PlacesList')).toHaveLength(0);
		const alert = wrapper.find('Alert');
		expect(alert).toHaveLength(1);
		expect(alert.text()).toEqual(mockErrorMessage);
		expect(wrapper).toMatchSnapshot();
	});

	it('renders SearchForm and Alert if there are no places found', () => {
		const alertMessage = 'There are no places related to your search';
		const props = { ...initialProps, places: [] };
		const wrapper = shallow(<App {...props} />);
		expect(wrapper.find('Spinner')).toHaveLength(0);
		expect(wrapper.find('SearchForm')).toHaveLength(1);
		expect(wrapper.find('PlacesList')).toHaveLength(0);
		const alert = wrapper.find('Alert');
		expect(alert).toHaveLength(1);
		expect(alert.text()).toEqual(alertMessage);
		expect(wrapper).toMatchSnapshot();
	});

	it('renders SearchForm and places list if there are places', () => {
		const props = {
			...initialProps,
			places: [
				{
					id: 'mock-id',
					provider: 'mock-provider',
					name: 'mock-name',
					address: 'mock-address',
				},
			],
		};
		const wrapper = shallow(<App {...props} />);
		expect(wrapper.find('Spinner')).toHaveLength(0);
		expect(wrapper.find('SearchForm')).toHaveLength(1);
		expect(wrapper.find('Alert')).toHaveLength(0);
		expect(wrapper.find('PlacesList')).toHaveLength(1);
		expect(wrapper).toMatchSnapshot();
	});
});
