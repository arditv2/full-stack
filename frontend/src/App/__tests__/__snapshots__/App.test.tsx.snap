// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders SearchForm and Alert if there are no places found 1`] = `
<Container
  className="mb-3"
  fluid={true}
>
  <Row
    noGutters={false}
  >
    <Col
      lg={
        Object {
          "offset": 3,
          "span": 6,
        }
      }
      md={
        Object {
          "offset": 2,
          "span": 8,
        }
      }
    >
      <SearchForm
        handleSubmit={[MockFunction]}
      />
      <Alert
        closeLabel="Close alert"
        show={true}
        transition={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "appear": false,
              "in": false,
              "mountOnEnter": false,
              "timeout": 300,
              "unmountOnExit": false,
            },
            "displayName": "Fade",
            "render": [Function],
          }
        }
        variant="info"
      >
        There are no places related to your search
      </Alert>
    </Col>
  </Row>
</Container>
`;

exports[`App component renders SearchForm and Alert if there is an error 1`] = `
<Container
  className="mb-3"
  fluid={true}
>
  <Row
    noGutters={false}
  >
    <Col
      lg={
        Object {
          "offset": 3,
          "span": 6,
        }
      }
      md={
        Object {
          "offset": 2,
          "span": 8,
        }
      }
    >
      <SearchForm
        handleSubmit={[MockFunction]}
      />
      <Alert
        closeLabel="Close alert"
        show={true}
        transition={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "appear": false,
              "in": false,
              "mountOnEnter": false,
              "timeout": 300,
              "unmountOnExit": false,
            },
            "displayName": "Fade",
            "render": [Function],
          }
        }
        variant="danger"
      >
        Something failed
      </Alert>
    </Col>
  </Row>
</Container>
`;

exports[`App component renders SearchForm and places list if there are places 1`] = `
<Container
  className="mb-3"
  fluid={true}
>
  <Row
    noGutters={false}
  >
    <Col
      lg={
        Object {
          "offset": 3,
          "span": 6,
        }
      }
      md={
        Object {
          "offset": 2,
          "span": 8,
        }
      }
    >
      <SearchForm
        handleSubmit={[MockFunction]}
      />
      <PlacesList
        places={
          Array [
            Object {
              "address": "mock-address",
              "id": "mock-id",
              "name": "mock-name",
              "provider": "mock-provider",
            },
          ]
        }
      />
    </Col>
  </Row>
</Container>
`;

exports[`App component renders only Spinner when loading 1`] = `
<Container
  className="mb-3"
  fluid={true}
>
  <Row
    noGutters={false}
  >
    <Col
      lg={
        Object {
          "offset": 3,
          "span": 6,
        }
      }
      md={
        Object {
          "offset": 2,
          "span": 8,
        }
      }
    >
      <Container
        className="spinner_container"
        fluid={true}
      >
        <Spinner
          animation="border"
          variant="primary"
        />
      </Container>
    </Col>
  </Row>
</Container>
`;
